### PR TITLE
Dn api 163 public identity methods

### PR DESCRIPTION
**Pull request by Dudleyneedham**
## fixes #123 
Started the API documentation for public identity methods.

Needs reviewing to check, for any spelling or grammar mistakes or any problems in the logic. 

## Checklist:

- [x] I have verified that the code works
- [x] I have verified that the code is easy to understand
  - [x] If not, I have left a well-balanced amount of inline comments
- [x] I have [left the code in a better state](https://deviq.com/boy-scout-rule/)
- [x] I have documented the changes (where applicable)

### CODE REVIEW

File: src/identity/Identity.ts [hunk 1]
```diff
@@ -155,6 +395,12 @@ export default class Identity extends PublicIdentity {
     )
   }
 
+  /**
+   * Signs a submittable extrinsic
+   * @param submittableExtrinsic
+   * @param nonceAsHex
```
Comment (maudnals) [inline src/identity/Identity.ts]: * Missing: example and parameters description

File: src/identity/Identity.ts [hunk 2]
```diff
@@ -28,9 +28,43 @@ type BoxPublicKey =
 
 export default class Identity extends PublicIdentity {
   private static ADDITIONAL_ENTROPY_FOR_HASHING = new Uint8Array([1, 2, 3])
+  /**
+   * [STATIC] Generates Mnemonic phrase used to create identities from phrase seed.
+   * @returns `generate`
```
Comment (maudnals) [inline src/identity/Identity.ts]: Fix: see what `generateMnemonic` returns in https://kiltprotocol.github.io/kilt-workshop-101/#/identity

File: src/identity/Identity.ts [hunk 3]
```diff
@@ -51,21 +85,89 @@ export default class Identity extends PublicIdentity {
...
+   * // testerIdentity.seed = seed: Uint8Array [
+   * // 108, 233, 253,  6,  12, 112,  22,  92,
+   * // 15, 200, 218, 37, 129,  13,  36, 145,
+   * // 6, 213, 223, 16,  10, 169, 128, 224,
+   * // 217, 161,  20,  9, 214, 179,  82,  97
+   * // ],
+   *
+   * const testerIdentityCheck = Kilt.Identity.buildFromSeed(tester.seed);
+   *
+   * // testerIdentityCheck.address 5HXfLqrqbKoKyi61YErwUrWEa1PWxikEojV7PCnLJgxrWd6W
```
Comment (maudnals) [inline src/identity/Identity.ts]: Suggestion: the example is here help users/us use the method properly. Part 1 is great but part 2 (check that the address the same and the seed is the same) is out of scope to me.

File: src/identity/Identity.ts [hunk 4]
```diff
@@ -51,21 +85,89 @@ export default class Identity extends PublicIdentity {
...
+   * ```
+   */
   public static buildFromSeed(seed: Uint8Array) {
     const keyring = new Keyring({ type: 'ed25519' })
     const keyringPair = keyring.addFromSeed(seed)
     return new Identity(seed, keyringPair)
   }
-
+  /**
+   * [STATIC] Builds a new Identity, generated from a uniform resource identifier (URIs).
```
Comment (maudnals) [inline src/identity/Identity.ts]: Typo: URI

File: src/identity/Identity.ts [hunk 5]
```diff
@@ -51,21 +85,89 @@ export default class Identity extends PublicIdentity {
...
+   * @param uri A phrase built from a mnemonic string.
+   * @returns `Identity`
+   * @example
+   * ```javascript
+   * // An Identity of Bob
```
Comment (maudnals) [inline src/identity/Identity.ts]: Suggestion: the comment here is extraneous, I think

File: src/identity/Identity.ts [hunk 6]
```diff
@@ -100,20 +202,76 @@ export default class Identity extends PublicIdentity {
...
 
   public getPublicIdentity(): PublicIdentity {
     const { address, boxPublicKeyAsHex } = this
     return { address, boxPublicKeyAsHex }
   }
 
+  /**
+   * Adds the signing for a message with an Identity using signing key ring pairs.
+   * @param cryptoInput
+   * @returns `Crypto`
```
Comment (maudnals) [inline src/identity/Identity.ts]: Fix: this returns the output of `Crypto.sign(...)`

File: src/identity/PublicIdentity.ts [hunk 7]
```diff
@@ -44,7 +51,12 @@ export default class PublicIdentity implements IPublicIdentity {
       return undefined
     }
   }
-
+  /**
+   * [STATIC] Resolve internal and external DIDs from a specific identifier
+   * @param identifier
```
Comment (maudnals) [inline src/identity/PublicIdentity.ts]: * Missing: example and parameters description

File: src/identity/Identity.ts [hunk 8]
```diff
@@ -28,9 +28,43 @@ type BoxPublicKey =
...
+   *
+   * @example
+   *
+   * ```javascript
+   * //Build a Mnemonic to use as an argument
+   *
+   * const mnemonic = Identity.generateMnemonic()
+   * // "mnemonic: coast ugly state lunch repeat step armed goose together pottery bind mention"
+   *
+   * const testIdentity = Identity.buildFromMnemonic(mnemonic)
```
Comment (maudnals) [inline src/identity/Identity.ts]: Great example!

File: src/identity/Identity.ts [hunk 9]
```diff
@@ -28,9 +28,43 @@ type BoxPublicKey =
...
+   * // testIdentity: Identity {
```
Comment (maudnals) [inline src/identity/Identity.ts]: Maybe here add "Output:", this will clarify what this comment is

File: src/identity/Identity.ts (hunk 2 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: I am returning the generated mnemonic. How do you mean?

File: src/identity/Identity.ts (hunk 8 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Ty

File: src/identity/Identity.ts (hunk 9 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: An output added, maybe like this?

File: src/identity/Identity.ts (hunk 3 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Swapped it out and removed the redundancy.

File: src/identity/Identity.ts (hunk 5 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Okie dokie

File: src/identity/Identity.ts [hunk 10]
```diff
@@ -100,20 +202,76 @@ export default class Identity extends PublicIdentity {
...
 
+  /**
+   * Encrypt symmetrical using random secret key (string)
+   * @param cryptoInput
+   * @param boxPublicKey
+   * @returns `Crypto`
+   * @example
+   * ```javascript
+   *  const secret =
+   * '0x000102030405060708090A0B0C0D0E0F' + '101112131415161718191A1B1C1D1E1F'
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Made the changes

File: src/identity/Identity.ts [hunk 11]
```diff
@@ -144,6 +361,29 @@ export default class Identity extends PublicIdentity {
     )
   }
 
+  /**
+   * Decrypt symmetrical using random secret key (UInt8Array)
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Made alternations

File: src/identity/Identity.ts [hunk 12]
```diff
@@ -155,6 +395,12 @@ export default class Identity extends PublicIdentity {
     )
   }
 
+  /**
+   * Signs a submittable extrinsic
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Added it

File: src/identity/Identity.ts [hunk 13]
```diff
@@ -100,20 +202,76 @@ export default class Identity extends PublicIdentity {
 
   private readonly signKeyringPair: KeyringPair
   private readonly boxKeyPair: BoxKeyPair
+  /**
+   * Creates a new instance of public identity.
+   * @returns `address`
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Have added.

File: src/identity/Identity.ts [hunk 14]
```diff
@@ -100,20 +202,76 @@ export default class Identity extends PublicIdentity {
...
+   * const signature = Crypto.sign(message, alice.signKeyringPair)
+   *
+   * ```
+   */
   public sign(cryptoInput: CryptoInput) {
     return Crypto.sign(cryptoInput, this.signKeyringPair)
   }
 
+  /**
+   * Adds the signing for a message string with an Identity using signing key ring pairs.
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Changed it

File: src/identity/Identity.ts (hunk 6 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Changed this out and updated the examples

File: src/identity/PublicIdentity.ts (hunk 7 above)
Comment (Dudleyneedham) [reply src/identity/PublicIdentity.ts]: Found this really difficult to find an example.

File: src/identity/PublicIdentity.ts [hunk 15]
```diff
@@ -15,6 +15,17 @@ export interface IURLResolver {
 }
 
 export default class PublicIdentity implements IPublicIdentity {
+  /**
+   * @description (STATIC) Creates a new Public Identity from a DID (Decentralised identifier) object.
+   * @param didDocument Public keys and pseudonymous biometrics
```
Comment (maudnals) [inline src/identity/PublicIdentity.ts]: Great that you added a definition, yet to me it doesn't describe what a didDocument is (also I'm not sure what biometrics have to do with DIDs, at least in the base case?).  
To see what a didDocument is, the IPublicIdentity test file and the official did spec might be good places to look.

File: src/identity/PublicIdentity.ts [hunk 16]
```diff
@@ -15,6 +15,17 @@ export interface IURLResolver {
...
+   * @returns A new [[PublicIdentity]] object
+   * @example
+   * ```javascript
+   *
+   * PublicIdentity.fromDidDocument(didDocument);
```
Comment (maudnals) [inline src/identity/PublicIdentity.ts]: `fromDidDocument` has no side effects, so this example is not realistic. Instead you can do: 
```
const identity = PublicIdentity.fromDidDocument(didDocument);
```

File: src/identity/Identity.ts [hunk 17]
```diff
@@ -100,20 +192,101 @@ export default class Identity extends PublicIdentity {
 
   private readonly signKeyringPair: KeyringPair
   private readonly boxKeyPair: BoxKeyPair
+  /**
+   * @description Creates a new instance of public identity.
+   * @returns An `address` which identity's a specific account on the chain.
```
Comment (maudnals) [inline src/identity/Identity.ts]: Not sure about the grammar 🙃

File: src/identity/Identity.ts [hunk 18]
```diff
@@ -100,20 +192,101 @@ export default class Identity extends PublicIdentity {
 
   private readonly signKeyringPair: KeyringPair
   private readonly boxKeyPair: BoxKeyPair
+  /**
+   * @description Creates a new instance of public identity.
```
Comment (maudnals) [inline src/identity/Identity.ts]: * `getPublicIdentity()` doesn't create a new instance. It returns the public identity
* Maybe explain why this is the public identity

File: src/identity/PublicIdentity.ts [hunk 19]
```diff
@@ -44,7 +55,18 @@ export default class PublicIdentity implements IPublicIdentity {
...
-
+  /**
+   * @description (STATIC) (ASYNC) Resolves internal and external DIDs (Decentralised identifier) from a specific identifier
+   * @param identifier URL scheme identifier
+   * @param urlResolver  DID URL always identifies the resource to be located
+   * @returns  A new [[PublicIdentity]] object
+   * @example
+   * ```javascript
+   *
+   * PublicIdentity.resolveFromDid(identifier, urlResolver)
```
Comment (maudnals) [inline src/identity/PublicIdentity.ts]: There should be an assignment (see my comment on `didDocument`).

File: src/identity/Identity.ts [hunk 20]
```diff
@@ -100,20 +192,101 @@ export default class Identity extends PublicIdentity {
...
+   * @returns An `address` which identity's a specific account on the chain.
+   * @returns  A `boxPublicKeyAsHex` public half of the cryptographic key pair in hexadecimal
+   * @example
+   * ```javascript
+   *
+   * const alice = Kilt.Identity.buildFromMnemonic();
+   *
+   * // The identity is able to call the getPublicIdentity()
+   *
+   * alice.getPublicIdentity()
```
Comment (maudnals) [inline src/identity/Identity.ts]: There should be an assignment (see my comment on `didDocument`).

File: src/identity/Identity.ts [hunk 21]
```diff
@@ -136,6 +331,38 @@ export default class Identity extends PublicIdentity {
     )
   }
 
+  /**
+   * @description Encrypts Asymmetrical using random secret key (UInt8Array)
```
Comment (maudnals) [inline src/identity/Identity.ts]: * Not sure about the grammar 🙃
* Maybe give some context about what the UInt8Array is (or if you prefer, don't mention this at all, the example might be sufficient)

File: src/identity/Identity.ts [hunk 22]
```diff
@@ -144,6 +371,30 @@ export default class Identity extends PublicIdentity {
     )
   }
 
+  /**
+   * @description Decrypts Asymmetrical using random secret key (UInt8Array)
```
Comment (maudnals) [inline src/identity/Identity.ts]: Not sure about the grammar 🙃

File: src/identity/Identity.ts [hunk 23]
```diff
@@ -144,6 +371,30 @@ export default class Identity extends PublicIdentity {
...
+   * @param encrypted An excrypted message/data
```
Comment (maudnals) [inline src/identity/Identity.ts]: Typo "encrypted"

File: src/identity/Identity.ts [hunk 24]
```diff
@@ -144,6 +371,30 @@ export default class Identity extends PublicIdentity {
...
+   * @param boxPublicKey A random secret key
+   * @returns `Crypto.decryptAsymmetric(...)`
```
Comment (maudnals) [inline src/identity/Identity.ts]: Let's document what this actually returns, aka what `Crypto.decryptAsymmetric()` returns.

File: src/identity/Identity.ts [hunk 25]
```diff
@@ -155,6 +406,19 @@ export default class Identity extends PublicIdentity {
     )
   }
 
+  /**
+   * @description Signs a submittable extrinsic, an extrinsic are pieces of information not from within the state of the chain.
```
Comment (maudnals) [inline src/identity/Identity.ts]: Not sure about the grammar 🙃

File: src/identity/Identity.ts [hunk 26]
```diff
@@ -155,6 +406,19 @@ export default class Identity extends PublicIdentity {
...
+   * @param submittableExtrinsic A transaction made to the chain.
+   * @param nonceAsHex a hex number added to a hashed or encrypted
```
Comment (maudnals) [inline src/identity/Identity.ts]: Missing a word?

File: src/identity/Identity.ts (hunk 18 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: I changed the wording and cleared it up a bit.

File: src/identity/Identity.ts (hunk 17 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Yes, you are right 👍 . I changed it and hopefully a little more clear.

File: src/identity/Identity.ts (hunk 23 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Cheers 👍

File: src/identity/PublicIdentity.ts (hunk 16 above)
Comment (Dudleyneedham) [reply src/identity/PublicIdentity.ts]: Nice, I have updated it.

File: src/identity/PublicIdentity.ts (hunk 19 above)
Comment (Dudleyneedham) [reply src/identity/PublicIdentity.ts]: I have updated it 👍

File: src/identity/Identity.ts (hunk 20 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Updated :)

File: src/identity/Identity.ts (hunk 21 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Updated and removed the Unit8Array.

File: src/identity/Identity.ts (hunk 25 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: I have updated, hopefully, should be better.

File: src/identity/Identity.ts (hunk 26 above)
Comment (Dudleyneedham) [reply src/identity/Identity.ts]: Updated, let me know if it is better.

File: src/identity/PublicIdentity.ts (hunk 15 above)
Comment (Dudleyneedham) [reply src/identity/PublicIdentity.ts]: changed it up, what do you think now?

**Pull request merged**